### PR TITLE
Plfm 4986 - datacite metadata to XML

### PR DESCRIPTION
**Pull request by nickgros-sage**
Create and test the DataciteMetadata object, an internal representation of the datacite metadata that can serialize the metadata into XML

### CODE REVIEW

File: lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadata.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	public void setResourceType(DoiResourceType resourceType) {
+		if (resourceType == null) {
+			throw new IllegalArgumentException("Missing general resource type. General resource type is required");
+		}
+		this.resourceType = resourceType;
+	}
+
+	public DoiResourceType getResourceType() {
+		if (resourceType == null) {
+			throw new NullPointerException("Missing general resource type. General resource type is required");
```
Comment (john-hill) [inline lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadata.java]: IllegalArgumentException

File: lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadataConstants.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	static final String CREATORS = "creators";
+	static final String TITLE = "title";
+	static final String TITLES = "titles";
+	static final String PUBLISHER = "publisher";
+	static final String PUBLICATION_YEAR = "publicationYear";
+	static final String RESOURCE_TYPE = "resourceType";
+	static final String RESOURCE_TYPE_GENERAL = "resourceTypeGeneral";
+
+	static String getSchemeUri(NameIdentifierSchemes scheme) {
+		String uri = null;
```
Comment (john-hill) [inline lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadataConstants.java]: add a test for both cases.

File: lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadataConstants.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+		switch (scheme) {
+			case ORCID:
+				uri = "http://orcid.org/";
+				break;
+			case ISNI:
+				uri = "http://www.insi.org/";
+				break;
+		}
```
Comment (john-hill) [inline lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadataConstants.java]: default: throw IllegalArgument("unknown type:")

File: lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadataConstants.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+		return uri;
+	}
```
Comment (john-hill) [inline lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadataConstants.java]: add a test that calls this for all types to protect when a new type is added, they  must also update this switch.

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DataciteDoi.json [hunk 5]
```diff
@@ -1,31 +1,37 @@
...
 			"type": "array",
 			"items": {
 				"$ref": "org.sagebionetworks.repo.model.doi.DoiTitle"
 			},
-			"required": true
+			"required": false
 		},
 		"publicationYear": {
 			"description": "The year that this resource became publicly accessible. Must be in YYYY format.",
-			"type": "integer",
```
Comment (nickgros-sage) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DataciteDoi.json]: keep integer

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DataciteDoi.json [hunk 6]
```diff
@@ -1,31 +1,37 @@
...
+			"required": true
+		},
 		"creators":{
 			"type":"array",
 			"description":"The main researchers involved in producing the data, or the authors of the publication, in priority order.",
 			"items":{
 				"$ref":"org.sagebionetworks.repo.model.doi.DoiCreator"
 			},
-			"required": true
+			"required": false
```
Comment (nickgros-sage) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DataciteDoi.json]: omit required

Comment (nickgros-sage) [reply lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DataciteDoi.json]: ALL of them

Comment (nickgros-sage) [reply lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DataciteDoi.json]: Note if parameter is optional/required in description

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json [hunk 7]
```diff
@@ -1,5 +1,11 @@
 {
-	"description": "JSON schema for DOI.",
```
Comment (nickgros-sage) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json]: rename file to DoiV2, leave current one alone

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json [hunk 8]
```diff
@@ -16,8 +22,13 @@
 			"$ref": "org.sagebionetworks.repo.model.doi.DoiStatus",
```
Comment (nickgros-sage) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json]: keep etag

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json [hunk 9]
```diff
@@ -1,5 +1,11 @@
...
-	"description": "JSON schema for DOI.",
+	"description": "JSON schema for all fields associated with a DOI and its metadata.",
+	"type":"object",
+	"implements": [
+		{
+			"$ref": "org.sagebionetworks.repo.model.doi.DataciteDoi"
+		}
+	],
 	"properties": {
 		"id": {
```
Comment (nickgros-sage) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json]: rename doiSymbol

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json (hunk 8 above)
Comment (nickgros-sage) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json]: remove doiStatus

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json [hunk 10]
```diff
@@ -16,8 +22,13 @@
...
+			"type": "string",
+			"description": "The URL that the DOI points to.",
+			"required": false
+		},
 		"objectId": {
-			"description": "The ID of the digital object for which this DOI is created.",
+			"description": "The ID of the digital object in Synapse for which this DOI is created.",
 			"type": "string",
 			"required": true
 		},
```
Comment (nickgros-sage) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json]: come up with a good name to distinguish Synapse metadata from DOI metadata (like creator)

Comment (nickgros-sage) [reply lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json]: userId
mintedOn
updatedOn
updatedBy

Comment (nickgros-sage) [reply lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/Doi.json]: mintedByUserId

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DoiId.json [hunk 11]
```diff
@@ -0,0 +1,26 @@
+{
+	"description": "JSON schema for all fields associated with a DOI and its metadata.",
+	"type":"object",
+	"properties": {
+		"id": {
```
Comment (nickgros-sage) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DoiId.json]: doiSymbol

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DoiId.json [hunk 12]
```diff
@@ -0,0 +1,26 @@
+{
+	"description": "JSON schema for all fields associated with a DOI and its metadata.",
+	"type":"object",
```
Comment (nickgros-sage) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DoiId.json]: convert to interface that has all of our fields (and have DoiV2 not create any new fields)

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DoiId.json [hunk 13]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (nickgros-sage) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/DoiId.json]: rename class to DoiAssociation

File: lib/lib-doi/src/test/java/org/sagebionetworks/doi/datacite/DataciteXmlTranslatorTest.java [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+	public void getResourceTypeTest() throws Exception {
+		Document dom = documentBuilder.parse(new InputSource(new StringReader("<resourceType resourceTypeGeneral=\"" + resourceType.getResourceTypeGeneral().name() + "\"></resourceType>")));
+		assertEquals(resourceType, getResourceType(dom));
+	}
+
+	@Test
+	public void testGetSchemeUri() {
+		for (NameIdentifierScheme e : NameIdentifierScheme.values()) {
+			String actual = getSchemeUri(e);
+			if (e == NameIdentifierScheme.ORCID) {
```
Comment (nickgros-sage) [inline lib/lib-doi/src/test/java/org/sagebionetworks/doi/datacite/DataciteXmlTranslatorTest.java]: just check that it's not null

Comment (nickgros-sage) [reply lib/lib-doi/src/test/java/org/sagebionetworks/doi/datacite/DataciteXmlTranslatorTest.java]: add another test case that tests that known schemes match the expected URI

File: lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadataTranslator.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package org.sagebionetworks.doi.datacite;
+
+import org.sagebionetworks.repo.model.doi.DataciteDoi;
+
+public interface DataciteMetadataTranslator {
```
Comment (nickgros-sage) [inline lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadataTranslator.java]: Add class comment to be more descriptive

Comment (nickgros-sage) [reply lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteMetadataTranslator.java]: explicitly state that class converts x to y

File: lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteXmlTranslatorImpl.java [hunk 16]
```diff
@@ -16,13 +16,31 @@
...
+		return translateUtil(xml);
+	}
 
 	/*
 	 * Translates a Datacite XML String into a DoiMetadata object
 	 * Works for XML adherent to Datacite Schemas 2.2 and 4.1
 	 */
-	public DoiMetadata translate(String xml) {
+	static Doi translateUtil(String xml) {
+		Document dom = parseXml(xml);
```
Comment (nickgros-sage) [inline lib/lib-doi/src/main/java/org/sagebionetworks/doi/datacite/DataciteXmlTranslatorImpl.java]: move parse up a method, then this method takes a document

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/v2/DoiAssociation.json [hunk 17]
```diff
@@ -0,0 +1,52 @@
+{
+	"description": "JSON schema for all fields associated with a DOI.",
+	"type":"interface",
+	"properties": {
+		"id": {
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/v2/DoiAssociation.json:5]: association id

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/v2/DoiAssociation.json [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+			"description": "The unique ID of this DOI stored in Synapse.",
+			"type": "string"
+		},
+		"etag": {
+			"type": "string",
+			"description": "For Optimistic Concurrency Control (OCC)."
+		},
+		"doiUri": {
+			"description": "The unique URI of this DOI to which the resource can be resolved. Contents of this field will be overwritten because it is determined by Synapse.",
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/v2/DoiAssociation.json:14]: Provided by Synapse.

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/v2/DoiAssociation.json [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+			"type": "string"
+		},
+		"doiUrl": {
+			"type": "string",
+			"description": "The DOI URL that will point to the Synapse object.  Contents of this field will be overwritten because it is determined by Synapse."
+		},
+		"objectId": {
```
Comment (john-hill) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/doi/v2/DoiAssociation.json:21]: required

**Pull request merged**